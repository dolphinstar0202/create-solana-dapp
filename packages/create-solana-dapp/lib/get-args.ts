import { log } from '@clack/prompts'
import { program } from 'commander'

import { CreateWorkspaceOptions } from 'create-nx-workspace'
import { anchorTemplates } from './anchor-templates'
import { getAppInfo } from './get-app-info'
import { GetArgsResult } from './get-args-result'
import { getPresets } from './get-presets'
import { getPrompts, getUiLibraries } from './get-prompts'

export async function getArgs(argv: string[]): Promise<GetArgsResult> {
  // Get app info from package.json
  const app = getAppInfo()

  // Get info about the presets
  const { presets, presetValues } = getPresets(app.version)

  // Get the result from the command line
  const input = program
    .name(app.name)
    .version(app.version)
    .argument('[name]', 'Name of the project (default: prompt)')
    .option(
      '-p, --preset <preset>',
      help(`Preset to use (default: prompt, options: ${presetValues.join(', ')})`),
      (value: string) => {
        if (!presetValues.includes(value)) {
          throw new Error(`Invalid preset: ${value}`)
        }
        return value
      },
    )
    .option('--ui <ui-library>', help(`UI library to use (default: prompt)`))
    .option(
      '-a, --anchor <template>',
      help(`Name of the Anchor template to use (default: prompt, set to false to disable)`),
      (value: string) => {
        if (!['false', ...anchorTemplates].includes(value)) {
          throw new Error(`Invalid anchor template: ${value}`)
        }
        return value
      },
    )
    .option('--app-name <name>', help(`Name of the frontend project (default: web)`))
    .option('--anchor-name <anchor-name>', help(`Anchor project name (default: anchor)`))
    .option('-pm, --package-manager <package-manager>', help(`Package manager to use (default: npm)`))
    .option('-d, --dry-run', 'Dry run (default: false)')
    .addHelpText(
      'after',
      `
Examples:
  $ ${app.name} my-app --preset react
  $ ${app.name} my-app --preset react --package-manager yarn
  $ ${app.name} my-app --preset react --anchor-template hello-world
      `,
    )
    .parse(argv)

  // Get the optional name argument (positional)
  const name = input.args[0]

  // Get the options from the command line
  const result = input.opts()

  // Take the result from the command line and use it to populate the options
  const options: GetArgsResult = {
    anchor: result.anchor,
    anchorName: result.anchorName ?? 'anchor',
    appName: result.appName ?? 'web',
    dryRun: result.dryRun ?? false,
    name: name ?? '',
    package: '',
    packageManager: (result.packageManager ?? 'npm') as CreateWorkspaceOptions['packageManager'],
    preset: result.preset,
    ui: result.ui,
  }

  // Get the prompts for any missing options
  const prompts = await getPrompts({ options, presets })

  // Populate the options with the prompts
  if (prompts.name) {
    options.name = prompts.name
  }
  if (prompts.preset) {
    options.preset = prompts.preset
  }
  if (prompts.ui) {
    options.ui = prompts.ui as string
  }
  if (prompts.anchor) {
    options.anchor = prompts.anchor
  }

  // Validate the options
  if (options.preset && !presetValues.includes(options.preset)) {
    log.error(`Invalid preset: ${options.preset}`)
    throw new Error(`Invalid preset: ${options.preset}`)
  }

  if (!['false', ...anchorTemplates].includes(options.anchor ?? '')) {
    log.error(`Invalid anchor template: ${options.anchor}`)
    throw new Error(`Invalid anchor template: ${options.anchor}`)
  }

  const libs = getUiLibraries(options.preset as 'next' | 'react')
  if (options.ui && !libs.map((l) => l.value).includes(options.ui)) {
    log.error(`Invalid ui library for preset ${options.preset}: ${options.ui}`)
    throw new Error(`Invalid ui library for preset ${options.preset}: ${options.ui}`)
  }

  // Create the package string
  // TODO: In the future, we should allow the user to specify the package and version using the preset option
  if (options.package === '') {
    const packageName = presets.find((preset) => preset.value === options.preset)?.package
    const packageVersion = presets.find((preset) => preset.value === options.preset)?.version
    options.package = `${packageName}@${packageVersion}`
  }

  return options
}

// Helper function to add a newline before the text
function help(text: string) {
  return `

  ${text}
`
}
