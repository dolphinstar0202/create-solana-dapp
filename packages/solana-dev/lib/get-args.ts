import * as p from '@clack/prompts'
import { Command } from 'commander'
import * as color from 'picocolors'
import { handleAnchorVersion } from './handle-anchor-version'
import { handleAvmVersion } from './handle-avm-version'
import { handleNodeVersion } from './handle-node-version'
import { handleRustVersion } from './handle-rust-version'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageJson = require('../package.json')

export const app = {
  name: packageJson.name,
  version: packageJson.version,
  supportUrl: 'https://solana.stackexchange.com/questions/tagged/solana-dev',
}

export async function getArgs(): Promise<void> {
  const program = new Command().name(app.name).version(app.version)

  program.addCommand(makeDoctorCommand())

  program.parse(process.argv)

  return
}

export function makeDoctorCommand() {
  return new Command('doctor').action(async () => {
    p.intro(`${color.bgCyan(color.black(` ${app.name} ${app.version} doctor `))}`)

    handleNodeVersion()
    handleRustVersion()
    handleAvmVersion()
    handleAnchorVersion()

    p.outro(`Problems? ${color.underline(color.cyan(app.supportUrl))}`)
  })
}
