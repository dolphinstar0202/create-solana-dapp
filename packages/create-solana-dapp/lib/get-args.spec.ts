import { getArgs } from './get-args'

describe('get-args', () => {
  describe('expected args', () => {
    it('should get args with minimal parameters', async () => {
      const result = await getArgs(['', '', 'my-app', '--preset', 'next', '--ui', 'tailwind', '--anchor', 'counter'])
      expect(result).toMatchSnapshot()
    })

    it.each([
      '--preset next --ui tailwind --anchor counter',
      '--preset react --ui tailwind --anchor counter',
      '--preset react --anchor none --ui none',
      '--preset react --anchor counter --anchor-name program --ui tailwind',
      '--preset=react --anchor=counter --anchor-name=program --ui=tailwind',
    ])('should get args with preset: %s', async (preset: string) => {
      const result = await getArgs(['', '', 'my-app', ...preset.split(' ')])

      expect(result).toMatchSnapshot()
    })
  })

  describe('unexpected args', () => {
    it('should fail with invalid preset', async () => {
      expect.assertions(1)
      try {
        await getArgs(['', '', 'my-app', '--preset', 'invalid'])
      } catch (e) {
        expect(e.message).toContain('Invalid preset: invalid')
      }
    })

    it('should fail with invalid anchor template', async () => {
      expect.assertions(1)
      try {
        await getArgs(['', '', 'my-app', '--preset', 'react', '--anchor', 'invalid'])
      } catch (e) {
        expect(e.message).toContain('Invalid anchor template: invalid')
      }
    })

    it('should fail with invalid ui library', async () => {
      expect.assertions(1)
      try {
        const x = await getArgs(['', '', 'my-app', '--preset', 'react', '--ui', 'jquery-ui'])
        console.log(x)
      } catch (e) {
        expect(e.message).toContain('Invalid ui library for preset react: jquery-ui')
      }
    })
  })
})
