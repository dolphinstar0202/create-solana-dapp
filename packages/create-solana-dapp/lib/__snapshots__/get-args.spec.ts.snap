// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-args expected args should get args with minimal parameters 1`] = `
{
  "anchor": "counter",
  "anchorName": "anchor",
  "appName": "web",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-next@0.0.1",
  "packageManager": "npm",
  "preset": "next",
  "ui": "tailwind",
}
`;

exports[`get-args expected args should get args with preset: --preset next --ui tailwind --anchor counter 1`] = `
{
  "anchor": "counter",
  "anchorName": "anchor",
  "appName": "web",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-next@0.0.1",
  "packageManager": "npm",
  "preset": "next",
  "ui": "tailwind",
}
`;

exports[`get-args expected args should get args with preset: --preset react --anchor counter --anchor-name program --ui tailwind 1`] = `
{
  "anchor": "counter",
  "anchorName": "program",
  "appName": "web",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "preset": "react",
  "ui": "tailwind",
}
`;

exports[`get-args expected args should get args with preset: --preset react --anchor none --ui none 1`] = `
{
  "anchor": "none",
  "anchorName": "anchor",
  "appName": "web",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "preset": "react",
  "ui": "none",
}
`;

exports[`get-args expected args should get args with preset: --preset react --ui tailwind --anchor counter 1`] = `
{
  "anchor": "counter",
  "anchorName": "anchor",
  "appName": "web",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "preset": "react",
  "ui": "tailwind",
}
`;

exports[`get-args expected args should get args with preset: --preset=react --anchor=counter --anchor-name=program --ui=tailwind 1`] = `
{
  "anchor": "counter",
  "anchorName": "program",
  "appName": "web",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "preset": "react",
  "ui": "tailwind",
}
`;
